### ISSUE TITLE
-Wdeclaration-after-statement in arm_neon.h

### ISSUE THREAD

**Issue by nathanchance** [llvm/llvm-project#54062]
The Linux kernel is looking to [upgrade from `-std=gnu89` to `-std=gnu11`](https://lore.kernel.org/r/CAHk-=wh97QY9fEQUK6zMVQwaQ_JWDvR=R+TxQ_0OYrMHQ+egvQ@mail.gmail.com/). The kernel builds with `-Wdeclaration-after-statement`, which exposes an issue in `arm_neon.h` with arm64 big endian: https://godbolt.org/z/313sW6zrW

```
$ cat repro.c
#include <arm_neon.h>

static inline uint8x16_t SHLBYTE(uint8x16_t v)
{
    return vshlq_n_u8(v, 1);
}

$ clang --target=aarch64-linux-gnu -mbig-endian -std=gnu11 -Wdeclaration-after-statement -fsyntax-only -ffreestanding repro.c
repro.c:5:9: warning: mixing declarations and code is incompatible with standards before C99 [-Wdeclaration-after-statement]
        return vshlq_n_u8(v, 1);
               ^
.../lib/clang/15.0.0/include/arm_neon.h:24823:14: note: expanded from macro 'vshlq_n_u8'
  uint8x16_t __ret; \
             ^
1 warning generated.

$ clang --target=aarch64-linux-gnu -mbig-endian -std=gnu89 -Wdeclaration-after-statement -fsyntax-only repro.c
```

As far as I understand based on reading `clang::ProcessWarningOptions()` and `DiagnosticIDs::getDiagnosticSeverity()`, there should not be any warnings from system headers if `-Wsystem-headers` is not passed, which the kernel does not do.

**Comment by DavidSpickett** [comment]
I don't think this issue is specific to this warning.

This "warning" changed recently:
https://reviews.llvm.org/D117232
https://reviews.llvm.org/rG118f966b46cfb60897b56a9878e1c68fd0e2afa4

Now when you are C99 or later it is a `Warning` vs. an `Extension`. (which means a warning or note about an extension from a later standard?)

The issue applies to other warnings too, at least coming from a macro in a system header. I haven't checked without the macro around them.

If I modify the little endian macro to add an unused variable and enable that warning I get:
```
$ ./bin/clang /tmp/test.c -c -o /tmp/test.o -Wunused -Wno-system-headers
/tmp/test.c:5:12: warning: unused variable 'a' [-Wunused-variable]
    return vshlq_n_u8(v, 1);
           ^
<...>build-llvm-aarch64/lib/clang/15.0.0/include/arm_neon.h:24814:7: note: expanded from macro 'vshlq_n_u8'
  int a; \
      ^
/tmp/test.c:3:26: warning: unused function 'SHLBYTE' [-Wunused-function]
static inline uint8x16_t SHLBYTE(uint8x16_t v)
                         ^
2 warnings generated.
```
I did a bit of printf hacking and it seems that clang is well aware that arm_neon.h is a system header and that the code comes from a macro within it (by using the "Spelling" location of the code the macro generates). However when it asks if the specific warning should be suppressed in system headers, the answer is no.

I thought perhaps it was a missing field in the way this new warning was added but I see very few explicitly put `SuppressInSystemHeader`/`SuppressInSystemMacro`. You can "fix" this by doing that but it seems like the wrong path.

So I think there's something going on where warnings in general aren't suppressed for system headers and/or macros within them.

**Comment by DavidSpickett** [comment]
Note to self (mostly), https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-system-header-prefix is an easier way to test this than hacking that header.

**Comment by DavidSpickett** [comment]
I think the macro is the differentiation here.

We have an unused var in a macro and in a function in this header:
```
$ cat foo/foo.h 
#define FOO int a; int b

inline void foo(void) {
  int z;
}
```

Then we include it and use both of them:
```
$ cat main.c
#include "foo.h"

int main() {
  FOO;
  foo();
  return 0;
}
```

Compile with `foo/` as a system header location, warn for unused, no system header warnings. We get a warning about the macro despite it originating in the system header foo.h.
```
$ ./bin/clang main.c -c -o main.o -Wunused -I./foo --system-header-prefix foo -Wno-system-headers
main.c:4:3: warning: unused variable 'a' [-Wunused-variable]
  FOO;
  ^
./foo/foo.h:1:17: note: expanded from macro 'FOO'
#define FOO int a; int b
                ^
main.c:4:3: warning: unused variable 'b' [-Wunused-variable]
./foo/foo.h:1:24: note: expanded from macro 'FOO'
#define FOO int a; int b
                       ^
2 warnings generated.
```
If we enable system header warnings we get warnings from the function and the macro.
```
$ ./bin/clang main.c -c -o main.o -Wunused -I./foo --system-header-prefix foo -Wsystem-headers
In file included from main.c:1:
./foo/foo.h:4:7: warning: unused variable 'z' [-Wunused-variable]
  int z;
      ^
main.c:4:3: warning: unused variable 'a' [-Wunused-variable]
  FOO;
  ^
./foo/foo.h:1:17: note: expanded from macro 'FOO'
#define FOO int a; int b
                ^
main.c:4:3: warning: unused variable 'b' [-Wunused-variable]
./foo/foo.h:1:24: note: expanded from macro 'FOO'
#define FOO int a; int b
                       ^
3 warnings generated.
```

If the warning from the system header function is ignored, so should the macro warning.

**Comment by nathanchance** [comment]
https://github.com/llvm/llvm-project/commit/c4db521cea32fcfb714d1a622e0efce69a564a28 mentions:

```
The current behavior is kept in this patch
(i.e. warnings from system macros are enabled by default).
The decision as to whether this should be an opt-in or opt-out
feature can be made in a separate patch.
```

[The review comments](https://reviews.llvm.org/D116833) make it seem like disabling warnings in macros from system headers is a little tricky so maybe it is better to just special case this?

**Comment by DavidSpickett** [comment]
Good find! Yes, this will fix this specific issue:
```
diff --git a/clang/include/clang/Basic/DiagnosticSemaKinds.td b/clang/include/clang/Basic/DiagnosticSemaKinds.td
index 15487260eb73..7a8204dd5648 100644
--- a/clang/include/clang/Basic/DiagnosticSemaKinds.td
+++ b/clang/include/clang/Basic/DiagnosticSemaKinds.td
@@ -9876,7 +9876,7 @@ def ext_mixed_decls_code : Extension<
   InGroup<DeclarationAfterStatement>;
 def warn_mixed_decls_code : Warning<
   "mixing declarations and code is incompatible with standards before C99">,
-  InGroup<DeclarationAfterStatement>, DefaultIgnore;
+  InGroup<DeclarationAfterStatement>, DefaultIgnore, SuppressInSystemMacro;

 def err_non_local_variable_decl_in_for : Error<
   "declaration of non-local variable in 'for' loop">;
```
However if https://reviews.llvm.org/D116833#3236209 is right and there is any chance of missing a legit warning elsewhere, we should change arm_neon.h to use the older style.

**Comment by nathanchance** [comment]
Right, I think that moving `arm_neon.h` to the older style is the lower risk change that would sufficiently resolve this from the kernel's perspective, as after https://git.kernel.org/linus/04e85bbf71c9072dcf0ad9a7150495d72461105c, it does not use system headers, aside from the few intrinsic headers.

**Comment by nickdesaulniers** [comment]
`Intrinsic::emitBody` in clang/utils/TableGen/NeonEmitter.cpp looks like that's what builds up the return variable declaration statemnet.

**Comment by nathanchance** [comment]
`Intrinsic::emitArgumentReversal()` [emits the declaration](https://github.com/llvm/llvm-project/blob/9e9c86fd6769a737d32b5f7a7513264d3cfc4f31/clang/utils/TableGen/NeonEmitter.cpp#L1225) and `Intrinsic::emitReverseVariable()` [emits the assignment](https://github.com/llvm/llvm-project/blob/9e9c86fd6769a737d32b5f7a7513264d3cfc4f31/clang/utils/TableGen/NeonEmitter.cpp#L1186-L1208).

I guess just combining those two functions so that the variable is both declared and assigned at the same time would be enough to silence the warning?

**Comment by nickdesaulniers** [comment]
1. https://reviews.llvm.org/D122188
2. https://reviews.llvm.org/D122189

**Comment by tstellar** [comment]
@nickdesaulniers Are there 2 commits for this or is it really just https://github.com/llvm/llvm-project/commit/5a2e56b70e2fa7ad0d82e54bc4c741b16f05e475 ?

**Comment by nickdesaulniers** [comment]
@tstellar it's just 5a2e56b70e2fa7ad0d82e54bc4c741b16f05e475;
D122188 (parent) was abandoned, D122189 (child) was updated.

**Comment by tstellar** [comment]
Merged: 63e7af77ec42de9ac0b2fca6f7876b6b67c40050

**Comment by ganelson** [comment]
These are difficult judgement calls, and I may not be seeing the whole picture here. But can I suggest that this well-intentioned change to clang seems to have done more harm than good?

The intention was that because the warning is off by default then it won't affect C99 users, but that isn't the case. If you compile a completely valid C99 program, and explicitly in C99 mode, but also specify -Weverything, then clang now issues just lots, and lots, and lots of declaration-after-statement warnings on any large C99 program, because -Weverything implicitly forces -Wdeclaration-after-statement; and declarations after statements are a core part of writing good idiomatic C99 code. So this change to clang came as a surprise to some clang users (it certainly surprised me), and has made a number of open-source projects have to change their makefiles.

It really seems to me that the original behaviour, i.e., that -Wdeclaration-after-statement did nothing in C99 mode, was correct. It strikes me as unfortunate to warn somebody coding in C99 that they're coding in C99; the new behaviour seems to exist so that users who want to write in a subset of C99 without such declarations can have clang do their linting for them. But we wouldn't expect clang to provide a warning for "you're using a union, not everybody likes unions", say. I don't really see how this is different from that.

**Comment by nickdesaulniers** [comment]
> If you compile a completely valid C99 program, and explicitly in C99 mode, but also specify -Weverything, then clang now issues just lots, and lots, and lots of declaration-after-statement warnings on any large C99 program, because -Weverything implicitly forces -Wdeclaration-after-statement

The intent of https://github.com/llvm/llvm-project/commit/5a2e56b70e2fa7ad0d82e54bc4c741b16f05e475 was to fix `-Wdeclaration-after-statement`. Can you provide an example of newly introduced instances of `-Wdeclaration-after-statement` as a result of https://github.com/llvm/llvm-project/commit/5a2e56b70e2fa7ad0d82e54bc4c741b16f05e475?

**Comment by ganelson** [comment]
Dear Nick, if I may,

So there's no dispute that clang is doing exactly what the fix for this bug intended - my suggestion is that the original "bug" was, in fact, not a bug but the correct behaviour. What I believe is now the case is that:

clang -std=c99 -c -Weverything

produces declaration-after-statement warnings every time any declaration is used after a statement. Probably most large C99 programs will exhibit this, but e.g., something like this:

#include <stdio.h>

int main() {
	printf("Hello\n");
	int i = 10;
	printf("i = %d\n", i);
	return 0;
}

Before the issue was resolved, I believe it did not throw a warning, and now it does. As a result we're changing all our makefiles to use -Weverything -Wno-declaration-after-statement instead of -Weverything, in order to suppress thousands of these warnings when we build, and of course that works. Some other open-source projects seem to have done the same. But I suspect this was not a consequence you intended.

To me it seems better to argue that declaration-after-statement warnings cannot ever be raised under -std=c99, because declaration after a statement is a normal part of C99 coding - in other words, had it been me, I would have closed the original issue as "behaviour is correct as it stands".

But the person who filed the original issue evidently felt the reverse. The trouble is that they were thinking of the use case where clang was invoked with an explicit -Wdeclaration-after-statement at the command line, so to them it seemed odd that no warnings then appeared. I don't agree, but I can see why they said it. What I'm saying is that changing the behaviour of -Wdeclaration-after-statement had the side-effect of changing the behaviour of -Weverything, so now we're at a point where -std=c99 -Weverything throws warnings which amount to saying that you're using a feature of C99. This seems to have caught at least some clang users by surprise.

But if I can just say - clang is a masterpiece, a really tremendous piece of work. On our little project (the programming language Inform), gcc can't even correctly compile or optimise our compiler. Clang has literally never failed us, and I generally think its diagnostics are the best I've ever seen in C compiler. Kudos.

  Graham Nelson


> On 22 May 2023, at 20:29, Nick Desaulniers ***@***.***> wrote:
> If you compile a completely valid C99 program, and explicitly in C99 mode, but also specify -Weverything, then clang now issues just lots, and lots, and lots of declaration-after-statement warnings on any large C99 program, because -Weverything implicitly forces -Wdeclaration-after-statement
> 
> The intent of 5a2e56b was to fix -Wdeclaration-after-statement. Can you provide an example of newly introduced instances of -Wdeclaration-after-statement as a result of 5a2e56b?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.
> 

--
Dr Graham Nelson
Fellow in Pure Mathematics, St Anne's College, Oxford

**Comment by nickdesaulniers** [comment]
>  but e.g., something like this:
> What I'm saying is that changing the behaviour of -Wdeclaration-after-statement had the side-effect of changing the behaviour of -Weverything, so now we're at a point where -std=c99 -Weverything throws warnings which amount to saying that you're using a feature of C99.

If you maintained a library that you expected to build cleanly as -std=c90 _and newer_, then I suspect you'd want -Wno-declaration-after-statement to warn regardless of language version.

---

Note that this particular thread is about ARM's arm_neon.h header; the behavior of -Wdeclaration-after-statement was not changed by https://github.com/llvm/llvm-project/commit/5a2e56b70e2fa7ad0d82e54bc4c741b16f05e475.

The behavior of -Wdeclaration-after-statement may have changed due some other change to clang, but this is the wrong thread to report such changes.  You may want to bisect clang, and find which change in particular caused any behavioral changes to -Wdeclaration-after-statement.

Also, we don't endorse the use of -Weverything. Some C++ warnings are non-mutually satisfiable.  This is documented policy: https://clang.llvm.org/docs/UsersManual.html#enabling-all-diagnostics . That said, if you have a codebase that builds diagnostic-free with -Weverything, congrats!

> But if I can just say - clang is a masterpiece, a really tremendous piece of work. On our little project (the programming language Inform), gcc can't even correctly compile or optimise our compiler. Clang has literally never failed us, and I generally think its diagnostics are the best I've ever seen in C compiler. Kudos.

That's generous; thank you!
